core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -0,0 +1,280 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import java.nio.charset.StandardCharsets
+import java.util
+import java.util.Properties
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+import kafka.log.LogConfig
+import kafka.server.DynamicBrokerConfig._
+import kafka.utils.{CoreUtils, Logging}
+import kafka.zk.{AdminZkClient, KafkaZkClient}
+import org.apache.kafka.common.Reconfigurable
+import org.apache.kafka.common.config.types.Password
+import org.apache.kafka.common.config.{ConfigDef, ConfigException, SslConfigs}
+import org.apache.kafka.common.network.{ChannelBuilder, ListenerReconfigurable}
+import org.apache.kafka.common.utils.Base64
+
+import scala.collection._
+import scala.collection.JavaConverters._
+
+object DynamicBrokerConfig {
```

Comment:
Some high level comments would be helpful here, in particular in regard to config precedence.